uefi/src/memory_descriptor.rs

```diff
@@ -0,0 +1,42 @@
+use bootloader_api::info::MemoryRegionKind;
+use bootloader_x86_64_common::legacy_memory_region::LegacyMemoryRegion;
+use uefi::table::boot::{MemoryDescriptor, MemoryType};
+use x86_64::PhysAddr;
+
+#[derive(Debug, Copy, Clone)]
+pub struct UefiMemoryDescriptor(pub MemoryDescriptor);
+
+const PAGE_SIZE: u64 = 4096;
+
+impl<'a> LegacyMemoryRegion for UefiMemoryDescriptor {
+    fn start(&self) -> PhysAddr {
+        PhysAddr::new(self.0.phys_start)
+    }
+
+    fn len(&self) -> u64 {
+        self.0.page_count * PAGE_SIZE
+    }
+
+    fn kind(&self) -> MemoryRegionKind {
+        match self.0.ty {
+            MemoryType::CONVENTIONAL => MemoryRegionKind::Usable,
+            other => MemoryRegionKind::UnknownUefi(other.0),
+        }
+    }
+
+    fn on_bootloader_exit(&mut self) {
+        match self.0.ty {
+            // the bootloader is about to exit, so we can reallocate its data
+            MemoryType::LOADER_CODE
+            | MemoryType::LOADER_DATA
+            | MemoryType::BOOT_SERVICES_CODE
+            | MemoryType::BOOT_SERVICES_DATA
+            | MemoryType::RUNTIME_SERVICES_CODE
+            | MemoryType::RUNTIME_SERVICES_DATA => {
+                // we don't need this data anymore
+                self.0.ty = MemoryType::CONVENTIONAL;
+            }
+            _ => {}
+        }
+    }
```

Comment:
This doesn't quite work because we also allocate the kernel's memory with `MemoryType::LOADER_DATA`:
https://github.com/rust-osdev/bootloader/blob/785e82a119d359edd93ef7fc9fc72a7e00f7c9da/uefi/src/main.rs#L201-L208

According to the UEFI spec the memory types `0x80000000..0xFFFFFFFF` are reserved for OS loaders. We could use of those for the kernel allocation.

Reply:
Good catch, thanks! Fixed in 73ae143c4395776ce32def52ba94d87a7671b967